lib/matplotlib/dates.py

```diff
@@ -740,12 +746,235 @@ def __call__(self, x, pos=0):
+                 zero_formats=None, show_offset=True):
+        """
+        Autoformat the date labels.  The default format is used to form an
+        initial string, and then redundant elements are removed.
+        """
+        self._locator = locator
+        self._tz = tz
+        self._oldticks = np.array([])
+        self._oldlabels = None
+        self.defaultfmt = '%Y'
+        # there are 6 levels with each level getting a specific format
+        # 0: mostly years,  1: months,  2: days,
+        # 3: hours, 4: minutes, 5: seconds
+        if formats:
+            if len(formats) != 6:
+                raise ValueError('formats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.formats = formats
+        else:
+            self.formats = ['%Y',  # ticks are mostly years
+                            '%b',          # ticks are mostly months
+                            '%d',          # ticks are mostly days
+                            '%H:%M',       # hrs
+                            '%H:%M',       # min
+                            '%S.%f',       # secs
+                            ]
+        # fmt for zeros ticks at this level.  These are
+        # ticks that should be labeled w/ info the level above.
+        # like 1 Jan can just be labled "Jan".  02:02:00 can
+        # just be labeled 02:02.
+        if zero_formats:
+            if len(formats) != 6:
+                raise ValueError('zero_formats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.zero_formats = zero_formats
+        elif formats:
+            # use the users formats for the zero tick formats
+            self.zero_formats = [''] + self.formats[:-1]
+        else:
+            # make the defaults a bit nicer:
+            self.zero_formats = [''] + self.formats[:-1]
+            self.zero_formats[3] = '%b-%d'
+
+        if offset_formats:
+            if len(offset_formats) != 6:
+                raise ValueError('offsetfmts argument must be a list of '
+                                 '6 format strings (or None)')
+            self.offset_formats = offset_formats
+        else:
+            self.offset_formats = ['',
+                                   '%Y',
+                                   '%Y-%b',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d',
+                                   '%Y-%b-%d %H:%M']
+        self.offset_string = ''
+        self._formatter = DateFormatter(self.defaultfmt, self._tz)
+        self.show_offset = show_offset
+
+    def __call__(self, x, pos=None):
+        if hasattr(self._locator, '_get_unit'):
+            locator_unit_scale = float(self._locator._get_unit())
+        else:
+            locator_unit_scale = 1.0
+        ticks = self._locator()
+        if pos is not None:
+            if not np.array_equal(ticks, self._oldticks):
```

Comment:
If I get this correctly, `_oldticks`/`_oldlabels` is an optimization to avoid recomputing all the ticks when the formatter gets called with pos=0, pos=1, pos=2, etc.
But that also means that if one assigns to `formats`/`offset_formats`/`zero_formats`, the labels don't get invalidated, so the assignment has no effect... until one changes the axes limits enough to make the ticks change.
I guess the options are either to make `formats`/... properties such that assigning to them invalidates `_oldlabels`, or also keep track of `_oldformats`, or use some helper that goes through `functools.lru_cache(1)`, or just not worry about the performance cost of doing the formatting multiple times.
(FWIW I think the formatter API is not optimal here; transforming tick positions to tick labels should be done as a single method call that takes all tick positions at the same time and returns all labels.)

Reply:
OK, it works w properties.  OTOH, the API is not what I'd have expected.  The "setter" doesn't get called if I do something like:

```python
formatter.format[0] = '%y'
```
and instead you must do:
```python
fmt = formatter.format
fmt[0] = '%y'
formatter.format = fmt
```
 
which I guess makes sense, but is not what I think would happen if it was just a bare class variable.

Reply:
Yeah, I failed to notice the problem with mutability, which means you need to track the old formats (in a local copy, as the lists can get modified in place...).
I agree with just redoing the formatting *n* times for now and change that if someone ever comes up with a use case where the performance cost matters...

The setter doesn't get called as `foo.bar[0] = baz` is equivalent to `tmp = foo.bar; tmp[0] = baz` so you call the getter, not the setter.

Reply:
I guess the other thing we could do is add `set_formats`, `get_formats` etc and make `self.formats` private.  Then it will be clear that you have to pass the whole list of formats.

Reply:
But that just seems to make the API more awkward to work around what may not even be a real performance issue.